pkg/compact/dedup/reader.go

```diff
@@ -0,0 +1,405 @@
+	if s.i >= len(s.samples) {
+		return false
+	}
+	s.i++
+	return true
+}
+
+func (s *SampleIterator) Seek(t int64) bool {
+	if s.i < 0 {
+		s.i = 0
+	}
+	for {
+		if s.i >= len(s.samples) {
+			return false
+		}
+		if s.samples[s.i].timestamp >= t {
+			return true
+		}
+		s.i++
+	}
+}
+
+type SampleSeries struct {
+	lset labels.Labels
+	data map[SampleType][]*Sample
+	res  int64
+}
+
+func NewSampleSeries(lset labels.Labels, data map[SampleType][]*Sample, res int64) *SampleSeries {
+	return &SampleSeries{
+		lset: lset,
+		data: data,
+		res:  res,
+	}
+}
+
+func (ss *SampleSeries) ToChunkSeries() (*ChunkSeries, error) {
+	if len(ss.data) == 0 {
+		return nil, nil
+	}
+	if ss.res == 0 {
+		return ss.toRawChunkSeries()
+	}
+	return ss.toDownsampleChunkSeries()
+}
+
+func (ss *SampleSeries) toRawChunkSeries() (*ChunkSeries, error) {
+	chks, err := ss.toChunks(RawSample)
+	if err != nil {
+		return nil, err
+	}
+	if len(chks) == 0 {
+		return nil, nil
+	}
+	return &ChunkSeries{
+		lset: ss.lset,
+		chks: chks,
+	}, nil
+}
+
+func (ss *SampleSeries) toDownsampleChunkSeries() (*ChunkSeries, error) {
+	all := make([][]chunks.Meta, len(downsampleAggrTypes), len(downsampleAggrTypes))
+	for _, at := range downsampleAggrTypes {
+		chks, err := ss.toChunks(typeMapping[at])
+		if err != nil {
+			return nil, err
+		}
+		all[at] = chks
+	}
+	numChks := len(all[downsample.AggrCount])
+	if numChks == 0 {
+		return nil, nil
+	}
+	result := make([]chunks.Meta, 0, numChks)
+	for i := 0; i < numChks; i++ {
+		var chks [5]chunkenc.Chunk
+		for _, at := range downsampleAggrTypes {
+			if all[at] == nil {
+				continue
+			}
+			chks[at] = all[at][i].Chunk
+		}
+		result = append(result, chunks.Meta{
+			MinTime: all[downsample.AggrCount][i].MinTime,
+			MaxTime: all[downsample.AggrCount][i].MaxTime,
+			Chunk:   downsample.EncodeAggrChunk(chks),
+		})
+	}
+
+	return &ChunkSeries{
+		lset: ss.lset,
+		chks: result,
+	}, nil
+}
+
+func (ss *SampleSeries) toChunks(st SampleType) ([]chunks.Meta, error) {
+	samples := ss.data[st]
+	if len(samples) == 0 {
+		return nil, nil
+	}
+	numChks := (len(samples)-1)/maxSamplesPerChunk + 1
+	chks := make([]chunks.Meta, 0, numChks)
+
+	for i := 0; i < numChks; i++ {
+		c := chunkenc.NewXORChunk()
+		appender, err := c.Appender()
+		if err != nil {
+			return nil, err
+		}
+		start := i * maxSamplesPerChunk
+		end := (i + 1) * maxSamplesPerChunk
+		if start == len(samples) {
+			break
+		}
+		if end > len(samples) {
+			end = len(samples)
+		}
+		for _, v := range samples[start:end] {
+			appender.Append(v.timestamp, v.value)
+		}
+		// InjectThanosMeta the chunk's counter aggregate with the last true sample.
+		if st == CounterSample {
+			appender.Append(samples[end-1].timestamp, samples[end-1].value)
+		}
+		chks = append(chks, chunks.Meta{
+			MinTime: samples[start].timestamp,
+			MaxTime: samples[end-1].timestamp,
+			Chunk:   c,
+		})
+	}
+	return chks, nil
+}
+
+type SampleReader struct {
```

Comment:
Hm. We deduplicate in `Iterators` interface currently... Wonder if we can reuse same algorithm as here: https://github.com/thanos-io/thanos/blob/master/pkg/query/iter.go#L375

Again - to reuse code to reduce overall complexity for understanding Thanos codebase (: and to gain benefits from improvements by changing just single place (:

Reply:
Actually this `SampleReader` doesn't do the deduplication logic. It just provides the function to read the samples under the specified time range.

Reply:
And using `DedupSeriesIterator` to do the dedup logic: https://github.com/thanos-io/thanos/blob/62cc9bf4fedd601209d8197daf9163ae484afae1/pkg/compact/dedup/merger.go#L421
So both of online and offline deduplication are using the same `penalty` algorithm.